### DISCUSSION TITLE
req removed from ws open handler

### DISCUSSION THREAD

**Discussion by acharkov** [uNetworking/uWebSockets.js#339] [Q&A] [answered]
I used _req_ from _open_ handler to get the real client address in case of proxy.
My old code was like this (at least in version 17.4.0)
```
open: (wsConn, req) => {
   ...
   this.ip = req.getHeader('cf-connecting-ip') || remoteAddressToString(conn.getRemoteAddress())
}
```
but in latest version _open_ handler is changed - _req_ removed. How I could get proxied ip for websocket connection in this case?

**Comment by hst-m** [comment, marked as answer]
v18 removes `req` from the ws open handler - [v18.0.0](https://github.com/uNetworking/uWebSockets.js/releases/tag/v18.0.0)
docs - [WebSocketBehavior.html#upgrade](https://unetworking.github.io/uWebSockets.js/generated/interfaces/WebSocketBehavior.html#upgrade)
docs - [HttpResponse.html#upgrade](https://unetworking.github.io/uWebSockets.js/generated/interfaces/HttpResponse.html#upgrade)
examples - [Upgrade.js](https://github.com/uNetworking/uWebSockets.js/blob/master/examples/Upgrade.js) - [UpgradeAsync.js](https://github.com/uNetworking/uWebSockets.js/blob/master/examples/UpgradeAsync.js)

you need to use the ws `upgrade` handler to pass values from the original http request:
```javascript
// pass values from ws upgrade handler

uws.App().ws('/', {
   upgrade: (res, req, context) => { // request was made to open websocket, res req have all properties for request, cookies etc
      // add code here to determine if ws request should be accepted or denied
      // deny request with "res.writeStatus('401').end()" see issue #367
      res.upgrade( // upgrade to websocket
         { ip: res.getRemoteAddressAsText() }, // 1st argument sets which properties to pass to ws object, in this case ip address
         req.getHeader('sec-websocket-key'),
         req.getHeader('sec-websocket-protocol'),
         req.getHeader('sec-websocket-extensions'), // 3 headers are used to setup websocket
         context // also used to setup websocket
      )
   },
   open: ws => {
      console.log(ws.ip)
   }
})
```

**Reply by wengjianhong** [reply]
docs - websocketbehavior.html#upgrade
docs - httpresponse.html#upgrade

can‘t request  now ??
